### PR TITLE
doc: update technical escalation proccess

### PR DESCRIPTION
**Pull request by nashif**
Update process for technical PR escalation.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

### CODE REVIEW

**Comment by kartben** [comment]
LGTM but this should probably be labelled TSC and discussed/approved by the TSC, right?

**Comment by nashif** [comment]
> LGTM but this should probably be labelled TSC and discussed/approved by the TSC, right?



> LGTM but this should probably be labelled TSC and discussed/approved by the TSC, right?

sure, the main thing I am adding here is the 1 business day requirement after dismissal which was not documented anywhere.

**Comment by kartben** [comment]
yup! To be clear I don't think there is anything controversial in there, but probably still worth the proper heads-up to folks.

File: doc/contribute/contributor_expectations.rst
Comment (aescolar) [inline doc/contribute/contributor_expectations.rst:230]: typo: 
.. is expected +to+ ensure..

File: doc/contribute/contributor_expectations.rst
Comment (henrikbrixandersen) [inline doc/contribute/contributor_expectations.rst:228]:
```suggestion
  binding resolution in the TSC by adding the `TSC`_ label on the PR.
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: doc/contribute/contributor_expectations.rst
Comment (aescolar) [reply doc/contribute/contributor_expectations.rst:225]: It was agreed long ago and approved in the TSC, that this was the case. That is, that a maintainer in an area can dismiss reviews from collaborators or people less involved in their area. And that in that case, those who disagree have the right to escalate to the TSC if they feel the need.
The idea is that maintainers are trusted to make decisions in their area to a large degree but an escalation path exists.
The situation otherwise was one in which anybody could block progress in an area even if they were not involved in it, and when a disagreement started, there was too much arm waving and other not proper behaviors that you'd expect from politicians and not engineers, which was not good at all. We had had some pointless confrontations before this rule, which this solved.
It is very rare that reviews are dismissed due to disagreements, and even rarer that they are escalated to the TSC.

Disenpowering maintainers and forcing any disagreement to be brought to a meeting in which most have no understanding of the issue would bring us back to a bad situation, so I would very much object to that. Those not involved in a review are typically those who have no stake or knowledge, so they should not be the default decision maker.

**Comment by keith-zephyr** [comment]
AIs:
@nashif - Will add language to require dismissals to provided justification.  Will also provide examples of "stale, irrelevant, and unrelated".  There also have to be discussion on the PR to agree on what is unrelated to the PR.

2023-10-10 Process WG notes:
@nashif This change to address how to unblock PRs.  Assignees already have the right to dismiss stale reviews.  This is meant to document the process for dismissing reviews and give reviewers a chance to respond.  Some reviewers believe that there review nack is always sticky.
@nashif we should also review the process for blocking changes.  "Request changes" is pretty low bar, and ends up blocking PRs.
@nashif also noted that dismissing a review can allow a PR to merge immediately, but we need a process to allow a review of the dismissal.
@dleach02 - Who has the rights to clear a block? Assignee is probably the best one.  The goal is that the normal PR process can clear blocks without having to dismiss reviews.  Any blockers put on by reviewers need a clear reason and a clear path to release the block.  How can we communicate the process to minimize these issues?   Is the dev group a better place to discuss this?
@fabiobaltieri - the recent PR that brought this to the forefront had problems on multiple fronts.
@nashif - dismissing reviews isn't an exception - is part of the defined process.   The issue is reviewers outside their maintained area adding blockers. but there isn't agreement by the area maintainer that the requested changes are the right path.
@dleach02 - the specific issue was a reviewer asking a PR to address tech debt outside the scope of the PR.  It's not good to stop progress on PRs.  This causes a negative experience to new contributors.
@fabiobaltieri - I think this is weaponizing the dismiss review process.   
@MaureenHelm - When is it appropriate to dismiss reviews?  
@dleach02 understands the concert about weaponizing dismissals.  
@fabiobaltieri - would like someone beside the author and reviewer to involved.
@nashif - these issues should never raise to the TSC for resolution.
@nashif - assignees have the responsibility to assign someone else if they are not the correct person.
@nashif - the rights of assignee to dismiss reviews is clearly stated in the documentation.
@dleach02 - what about assignee right to block PR from being merged?  @nashif - this is would be a case that should be escalated to the arch WG or dev WG.
@nashif - We need to define a block vs "request for changes".  
@MaureenHelm - This wasn't a stale review - the reviewer continually actively blocked the request because their concerns were not getting addressed.
@nashif - feature creep changes are good candidates for dismissal.  
@MaureenHelm - need to distinguish between stale reviews and blocking changes from an engaged reviewer
@fabiobaltieri - the specific PR was missing actionable feedback on the blocking review
@fabiobaltieri - Would like to avoid dismissing engaged reviews.
@keith-zephyr - suggests any dismissal needs to provide justification on the PR for the dismissmal.  This should point to the documentation.
@nashif - feature creep is a real issue to be addressed.  If requesting significant refactoring, reviewer should propose an RFC to fix.
@wilfried - What's the dismiss review process?  @dleach02 - the dismiss review clears the block from the other reviewer.
@dleach02 - agrees feature creep slows things down, especially when there isn't a consensus among the project for the direction.
@nashif - There may be too many paths/options for discussion and escalation.  It's not always clear if things should go to dev review or arch review.
@MaureenHelm - the dev review agenda is published in advance.  @dleach02 - normally would ping reviewer on discord
@nashif - need to ensure people involved can attend the dev/arch meeting when the issue discussed
@nashif - the documentation for dismissal is in 2 separate places.  The proposed change is a hotfix to codify the timeline for dismissal.
@nashif - will also look into documenting the expectations when adding a blocking review.
@fabiobaltieri  - still doesn't like given this power to the assignee.  @nashif assignee should be a maintainer and already has the power to set direction.
@nashif - thinks dismissal should be used more to keep PRs moving.
@keith-zephyr - reaching out on discord to discuss can be effective, but when the conversation can occur on the PR, it serves to document the full discussion.

File: doc/contribute/contributor_expectations.rst
Comment (nashif) [reply doc/contribute/contributor_expectations.rst:225]: > I think cases that result in review dismissals should be carried on by someone unrelated to the author and reviewers.

this is exactly what the escalation process is about.

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:225]: I like the idea that maintainer are trusted and empowered to make decisions, but I think that there's few corner cases that should be considered, for instance:
- Driver changes usually falls between two (or more) maintainer areas: the subsystem and the platform, who end up assigned is pretty arbitrary, IIRC it's based on the file count but patform has priority. I think in case of disagreement the subsystem maintainer should have priority over the platform one.
- Situations where the assignee author and in the same org, the reviewer is external. It may be alright to dismiss but in principle it does not really play well with the vendor neutrality nature of the project.

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:225]: > this is exactly what the escalation process is about.

Oh you mean up to arch or tsc? Sure but then the dismissal could be up to the working group.

Comment (nashif) [reply doc/contribute/contributor_expectations.rst:225]: > Oh you mean up to arch or tsc? Sure but then the dismissal could be up to the working group.

no, because you might dismiss some review that is blocking a PR for no good reason and the reviewer concerned might not even care or they might just accept it and move on, we do not want to make every disagreement go to a WG, the WG or TSC path only needed when disagreement continues and no way to resolve in the PR.

Comment (nashif) [reply doc/contribute/contributor_expectations.rst:225]: > * Driver changes usually falls between two (or more) maintainer areas: the subsystem and the platform, who end up assigned is pretty arbitrary, IIRC it's based on the file count but patform has priority. I think in case of disagreement the subsystem maintainer should have priority over the platform one.

we already document that maintainer have the responsibility to re-assign. Anyone can contest wrong assignment and re-assign if it makes sense. But we also have to expect that 'maintainer' should be looking for the project and the common good by default and that maintainers will do the right thing at the end. There is no system that is perfect and we need to have some trust. 

> Situations where the assignee author and in the same org, the reviewer is external. It may be alright to dismiss but in principle it does not really play well with the vendor neutrality nature of the project.

This is the reality of the project and we have recently added the measure to at least not allow merges of PRs when submitter, reviewers are from the same org, so we have at least one pair of eyes that is not involved directly. Any other reviews from other collaborators are still and will always be welcome, but the reality here is that the vendors know their hardware best. As long as the context of a change is "local" to one vendor, wrong doing will only harm the vendor themselves. 
This become more critical for "common" areas, and we already request that more eyes from different organisation are part of the review.

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:225]: > [...] but the reality here is that the vendors know their hardware best. As long as the context of a change is "local" to one vendor, wrong doing will only harm the vendor themselves.

That sounds like a situation where having external contributions would improve the overall result. A feedback from someone unfamiliar with the specific hardware may highlight documentation gaps that would otherwise be missed. Consciously accepting harmful changes just because it's withing a vendor set of drivers also does not strike me as ideal.

Anyway, I think I may be misreading this, is the intended sequence like:
- assignee dismisses the review
- the reviewer has 1 day to follow up

or
- the assignee asks the reviewer to clarify or escalate to arch
- the review gets dismissed if no response after 1 day

?

Comment (marc-hb) [reply doc/contribute/contributor_expectations.rst:225]: > That sounds like a situation where having external contributions would improve the overall result.

Yes of course and sometimes it happens which is unbelievable but this is realistically the situation where there is the least money and least interest to perform "outsider" reviews.

> Consciously accepting harmful changes...

Strike "consciously".

Comment (nashif) [reply doc/contribute/contributor_expectations.rst:225]: > Anyway, I think I may be misreading this, is the intended sequence like:
> 
> * assignee dismisses the review
> * the reviewer has 1 day to follow up

this.

There are many other steps and things that can be done before a review is dismissed, i.e. discussion in the PR itself, discord, the dev-review meeting etc or even the arch. WG if the issue at hand is at the level we want to have in this meeting, but ultimately and if all options have been considered and tried, the assignee can dismiss with a justification and eventually go through the escalation process if requested.

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:225]: > There are many other steps and things that can be done before a review is dismissed, i.e. discussion in the PR itself, discord, the dev-review meeting etc or even the arch. WG if the issue at hand is at the level we want to have in this meeting, but ultimately and if all options have been considered and tried, the assignee can dismiss with a justification and eventually go through the escalation process if requested.

Ok, I missed out that this was the last item in the escalation list and already after the "talk to each other" and "talk to each other in a working group" steps. Yeah I think this is fine then, though it's odd that the working group item is "optionally resolve...". I think a reviewer should  be given the chance to ask to bring up a PR in a working group before their review is dismissed. Maybe drop "optionally" from the previous step?

Comment (dleach02) [reply doc/contribute/contributor_expectations.rst:225]: Everything LGTM, but to @fabiobaltieri concerns, this section actually starts with resolving issues in the PR, then it says "optionally" take the conversation to the dev or arch meetings. Might I suggest that we remove the word "optionally" and just make the steps:

1. Work within the PR and community infrastructure (discord) to resolve the PR
2. Bring to the dev/arch meeting to discuss

Let the discussion results from the dev/arch be the trigger to decide if the review should be dismissed or not with the assignee being able to do the dismissal?

Comment (nashif) [reply doc/contribute/contributor_expectations.rst]: dev-review meeting is not part of the escalation, it is just a discussion forum and artch WG meeting should not be used to discuss every single disagreement in PRs, the backlog of technical discussion of new subsystems APIs etc is already long enough, this just does not scale, also, not everyone can attend those meetings. So escalation path should be short we defined and should not involve 3 meetings (dev-review, arch wg, TSC...).

Anyways, I added more updates, please take another look.

File: doc/contribute/contributor_expectations.rst
Comment (dleach02) [inline doc/contribute/contributor_expectations.rst:226]: The minimum of 1 business day to respond is by responding in the PR I believe but not sure how to tie this with the "or" escalate to the arch WG? Should this be, 1 day to respond and if objecting, escalate to the next arch WG meeting?

File: doc/contribute/contributor_expectations.rst
Comment (aescolar) [inline doc/contribute/contributor_expectations.rst:216]: Please leave the "optionally" (and ideally the whole paragraph as it was). As you @nashif said not everybody can join these meetings. We don't all live in US time. It is hardly helpful to demand that 2 people in Europe, India or Australia need to join a meeting in US pacific time in which things may or may not be discussed, in which there is high likelihood nobody else will care about the issue.
Having the dev-review and architecture meeting as an option/possibility for discussion is good. Forcing it as the path, no.

File: doc/contribute/contributor_expectations.rst
Comment (aescolar) [inline doc/contribute/contributor_expectations.rst:225]: Why the tripple escalation of dev-review, architecture working group and then TSC? (This seems like a lot of steps to keep dragging matters forever).
I would leave this as it was before (i.e. optionally discuss it in the dev-review or arch workgroup, then TSC).

**Review by aescolar — DISMISSED** [review]
Please revert the lack of optionality of the dev-review & arch meeting.
We don't all leave in America ;).

File: doc/contribute/contributor_expectations.rst
Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst]: Thanks I think it's a bit more clear now.

File: doc/contribute/contributor_expectations.rst
Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:226]: Yeah I also find the dynamics of this a bit awkward, how does merging plays into this? Review dismissed, should merging wait some time for the person to escalate?

Comment (nashif) [reply doc/contribute/contributor_expectations.rst]: addressed....

File: doc/contribute/contributor_expectations.rst
Comment (aescolar) [inline doc/contribute/contributor_expectations.rst:240]: Can we make this architecture working discussion optional together with the Zephyr Dev Meeting? so we don't have both the arch working group and the TSC?

Comment (nashif) [reply doc/contribute/contributor_expectations.rst:240]: I prefer to have this resolved much earlier and not having to deal with PR conflicrt resolution in the TSC. The TSC is the  very very last resort and should not be seen as the first thing you can do. I also do not see the Dev Meeting on the path for escalation, it is a discussion forum and it will be less confusing if we had less choices and less "or"s IMO.

File: doc/contribute/contributor_expectations.rst
Comment (henrikbrixandersen) [inline doc/contribute/contributor_expectations.rst:238]: Escalation by a the person who initiated the PR should also be allowed.

**Review by henrikbrixandersen — COMMENTED** [review]
Minor change suggestion, but looks good.

**Review by dleach02 — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by MaureenHelm — DISMISSED** [review]

File: doc/contribute/contributor_expectations.rst
Comment (keith-zephyr) [inline doc/contribute/contributor_expectations.rst:235]: There isn't any enforcement that assignees have approved a PR, correct?  I would simplify this language to just apply the DNM label.

File: doc/contribute/contributor_expectations.rst
Comment (keith-zephyr) [inline doc/contribute/contributor_expectations.rst:248]: You mention in a comment that the TSC is a last resort.  How about making that explicit here? 
```suggestion
- If all avenues of resolution and escalation have failed, assignees can escalate 
  to the TSC and get a binding resolution in the TSC by adding the *TSC* label
  on the PR.
```

File: doc/project/project_roles.rst
Comment (keith-zephyr) [inline doc/project/project_roles.rst:104]:
```suggestion
Collaborator approval of pull requests are counted toward the minimum

```

File: doc/project/project_roles.rst
Comment (keith-zephyr) [inline doc/project/project_roles.rst:131]:
```suggestion
Maintainer approval of pull requests are counted toward the minimum
```

File: doc/contribute/contributor_expectations.rst
Comment (rodrigopex) [inline doc/contribute/contributor_expectations.rst:249]: Far from being polemic, I believe we can solve all the disagreements earlier in the regular AG meeting. However, sometimes it can happen, and TSC's decision is vital. Is there any process regarding that? Will the TSC decide based on what information? Should the involved parties invoked to tell their point of view or something similar? 

Based on my situation as an independent maintainer. I won't have any direct representatives (as a company or association) on the TSC members. For me, It would be harder to succeed in a disagreement with a regular employee of a Zephyr platinum-member company, for example. Is that the intention, or are we trying to be more democratic?

Comment (jfischer-no) [reply doc/contribute/contributor_expectations.rst:249]: There are two community members in TSC.

Comment (rodrigopex) [reply doc/contribute/contributor_expectations.rst:249]: > There are two community members in TSC.

Thanks. Great to know. I tried to see who are they, but I could not.

Comment (henrikbrixandersen) [reply doc/contribute/contributor_expectations.rst:249]: Me and @erwango.

Comment (jfischer-no) [reply doc/contribute/contributor_expectations.rst:249]: There is even c in TSC for community, and small _n_ somewhere for neutrality.

File: doc/project/project_roles.rst
Comment (keith-zephyr) [inline doc/project/project_roles.rst:105]: Extra line.

File: doc/contribute/contributor_expectations.rst
Comment (mbolivar-ampere) [inline doc/contribute/contributor_expectations.rst:233]: s/escalation/escalate/

**Review by mbolivar-ampere — COMMENTED** [review]
Seems reasonable. In addition to other comments people already made, I saw one typo

**Review by kartben — APPROVED** [review]

**Review by aescolar — APPROVED** [review]

**Review by simhein — APPROVED** [review]

**Pull request merged**
